### DISCUSSION TITLE
Wrappers for passing multiple survey models to mice::pool

### DISCUSSION THREAD

**Discussion by bgall** [amices/mice#397] [Show and tell]
`mice::pool` takes a list of model objects and applies Rubin's rules using the point estimates, standard errors, and model residual degrees of freedom ("residualdf") to pool estimates across statistical models. Unfortunately, it can have issues when working with models/objects from the `survey` package.

`pool` relies on `broom::glance` to extract residual degrees of freedom from objects and `broom::tidy` to extract coefficients and standard errors, but `broom::glance` does not behave well when mapped to `survey` objects. For example, supposing a list of `survey` models `svy_obj`, `lapply(svy_obj, function(x) broom::glance(x))` will throw an error. See [here](https://github.com/tidymodels/broom/issues/1022) for detailed discussion of this issue, although note that this only happens when mapping `broom::glance` to a list and passing a single object poses no problem.

Although `broom::tidy` works with `survey` models and `mice::pool` successfully extracts information with this function, we still typically need the residual for inferences if we are interested in quantities such as p-values. Luckily, `survey` model objects do contain this quantity and `mice::pool` allows users to specify this quantity in lieu of using `broom::glance`. Accordingly, we can define a function to  extract the residualdf from the models and pass it to `pool` to produce the output expected when `pool` has all of the information it wants.

This may all be based on a misunderstanding of how to create a list of `survey` model objects and pass them to `mice:pool` (it would be great to see if @stefvanbuuren has a better handle on this), but if not, I put together provide a simple workaround that can be applied to survey models. See below.

``` r
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Dependencies
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

library(dplyr)
library(purrr)
library(broom)
library(mice)
library(survey)

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Define: svyglm_pool
# Takes a list of `survey` objects and a
# string formula to produce estimates using
# using Rubin's rules. Could substitute
# any `survey` model in lieu of `svyglm`
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

svyglm_pool <- function(svy_list, formula) {
  # Convert string to formula
  myform <- formula %>% as.formula()
  
  # Fit models
  mods <-
    purrr::map(svy_list, function(x) {
      svyglm(formula = myform, design = x)
    })
  
  # Grab residual degrees of freedom from
  # first model since same across all models
  # and `broom::glance()` currently doesn't
  # detect this.
  residual_df <- mods[[1]]$df.residual
  
  # Get pooled estimates across models
  mice::pool(mods, dfcom = residual_df)
  
}

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Examples
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

# Create example "imputed" data
df_list <- data.frame(z = rnorm(20),
                      y = rnorm(20)) %>%
  list() %>%
  rep(2)

# Turn data.frames into `survey.design` objects
# for use with `survey` model functions
objs_survey <- df_list %>%
  purrr::map(function(x) {
    survey::svydesign(data = x,
                      ids = ~ 1,
                      weights = ~ 1)
  })

# Fit linear models and get pooled estimates
objs_survey %>%
  svyglm_pool(formula = "y ~ z") %>%
  summary()
#>          term   estimate std.error  statistic       df   p.value
#> 1 (Intercept)  0.1352633 0.2015998  0.6709494 16.28408 0.5116582
#> 2           z -0.1088611 0.1674164 -0.6502415 16.28408 0.5246047
```

<sup>Created on 2021-05-19 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

**Comment by stefvanbuuren** [comment]
Thanks. If I understand correctly, the problem is that the `broom` package does not return all needed summary statistics from a `survey` model. 

Your work-around does model fitting and pooling as one function by combines the functionality of `mice::with.mids()` and `mice::pool()`. This makes your solution specific to `survey::svyglm()` model. 

I wondered about a slightly more generic approach: Run `fit <- mice::with.mids()` as usual, manually extract `residual_df` from `fit`, and then run `mice::pool(fit, dfcom = residual_df)`. I expect that would yield the same result.
